pkg/dio/multi.go

```diff
@@ -191,20 +205,25 @@ func (t *multiReader) Read(p []byte) (int, error) {
 
 		x, err := r.Read(slice)
 		n += x
-		// if any of the readers don't return EOF, it's not EOF
-		if err == nil {
-			eof = nil
-		} else if err != io.EOF {
-			// if there was an actual error, return that
-			// we cannot handle multiple not EOF errors, so return now
-			t.err = err
-			return n, err
+		if err != nil {
+			if err == io.EOF {
+				// increment the EOF counter and remove the reader that retured EOF
+				log.Debugf("[%p] returned EOF", r)
+				t.Remove(r)
+				eof++
+			} else {
+				// if there was an actual error, return that
+				t.err = err
+				return n, err
+			}
 		}
 	}
 
-	// we'd have returned anything other than EOF/nil inline
-	t.err = eof
-	return n, eof
+	// This means readers closed/removed while we iterate
+	if n == 0 && t.err == nil && eof == len(rTmp) {
```

Comment:
`if eof != 0 && ....` - otherwise we'll get this even when no eof was returned every time we remove the last writer.